### PR TITLE
Remove version sensor

### PR DESCRIPTION
**Pull request by richibrics**
Related to #92

### CODE REVIEW

**Comment by richibrics** [comment]
Current idea:
Entity that has update data can have
- current version sensor
- latest version sensor
- is new version available sensor
- update command

The problem is that each entity data has its own key and if I have the HomeAssistant warehouse that has to map these 4 data in one Update Entity in Hass, it doesn't know which data is what (e.g. doesn't know current version sensor needs to go in the configuration as the current version).

Two ideas:
1. Have fixed keys for the entity data to use so that the Hass WH will look for those to place in the Update Entity configurations in the correct spots.
2. (Favorite) Have the Entity data (those 4) in my Entity, then in order for the WH to know which data is the one for example of current version, we place in the entities.yaml the key of that data for the current version.

Example of entities.yaml for choice 2:
```
IoTuringUpdate:
  icon: mdi:update
  custom_type: update
  sensor_current_version: "THE KEY OF THE SENSOR IN MY ENTITY"
  command_update: "THE KEY OF THE COMMAND IN MY ENTITY"
  sensor_latest_version: "KEY..."
  sensor_is_version_available: "KEY..."
```

What do you think ? I don't know if I wrote it down in an understable way

**Comment by infeeeee** [comment]
Yeah, it's hard to follow what you want :)

My ideal would be something like this, how it would look like in HA:
- AppInfo sensor:
  - main attribute name and version: `IoTuring 2024.3.1`
  - extra attributes: installed version, latest version, commit hashes
- Update available binary sensor
- Update and Restart button, if no update available it simply restarts IoTuring

> The problem is that each entity data has its own key and if I have the HomeAssistant warehouse that has to map these 4 data in one Update Entity in Hass

I think the full update system could be rethinked. It would be useful if some sensors could have different update frequency. A really easy implementation would be if slow updating sensors would simply miss some loops, e.g. they update only every 10th loop.

> it doesn't know which data is what (e.g. doesn't know current version sensor needs to go in the configuration as the current version).

I don't understand what do you mean here.

>  we place in the entities.yaml

I don't think it would be useful, I think entities.yaml should be merged with configuration. That way users could change the icons, names of entities much more easily. What about other Whs, they don't use the yaml, but the keys are universal.

**Comment by richibrics** [comment]
Yeah sorry let me rewrite it.
Goal: have a [HomeAssistant "Update" entity](https://www.home-assistant.io/integrations/update.mqtt/) with IoTuring current version and available version with the future possibility of updating it from Hass.

In order to have the HassUpdate entity, we need to specify in the configuration we send via MQTT to Hass:
- the topic with the current version
- the topic with the latest version
- the topic to which send a message to start the installation (for future uses)

In order to get this behavior, I need my IoTuring entity to have two sensors that send the current and latest version and a command to start the installation.
Then in the Hass WH I want to translate the 2 sensors and the command to 1 entity in HASS (an update entity).
In order to do it, I need to provide in the MQTT configurations the topics of the sensors and of the commands in the corrects spots:
- topic of the sensor with current version in "state_topic" key in config
- topic of the sensor with latest version in "latest_version_topic" in config
- and the same for the command.

The problem is that the Hass WH doesn't know which of the sensors of my entity is the one with the current version and the one with the latest version.
To do it I thought we could provide to the WH the keys of the sensors of each information in the entities.yaml like I wrote above.

In this way when we type in the YAML that we want to configure an Update entity in Hass, then we place in the MQTT configurations the topics of the sensors I can correctly identify by using the keys I have in entities.yaml

e.g.
```
IoTuringUpdate:
  icon: mdi:update
  custom_type: update
  sensor_current_version: "THE KEY OF THE SENSOR IN MY ENTITY" -> this sensor's topic will be taken from the WH and will be placed in the MQTT configurations as "state_topic"
  .....
```

I hope it's better now; with Update I was not talking of the entities update interval

**Comment by infeeeee** [comment]
Ah, ok , I understand now, I didn't know about the MQTT Update entity.

Basically it's an `EntityCommand` with 2 connected `EntitySensors`, and one of the sensors uses `latest_version_topic` instead of the `state_topic`.

I would do it this way:
- Modify `EntityCommand` to allow 2 connected `EntitySensors`
- Create a subclass of `HomeAssistantEntityBase`, something similar to `LwtSensor`, which will send the `latest_version_topic`. Lets call it `SecondarySensor`.
- In `CollectEntityData` selects this new subclass for every additional connected sensors, first connected sensor remains the same.
- For discovery, `SecondarySensor` only needs this custom topic everything else can be  read from the first sensor. The topic could be read from the yaml:

```yaml
IoTuringUpdate:
  icon: mdi:update
  custom_type: update
  secondary_topic: latest_version_topic
```

- As it's still an `EntitySensor`, it updates the same way as any other sensors. Also this sensor is still available to other warehouses

**Comment by richibrics** [comment]
But this would be a problem if we would like to use the MQTT update entity without the command to trigger the update, since in IoTuring the command would become necessary isn't it ?

**Comment by infeeeee** [comment]
Good point, then the other way: create a sensor with 2 topics. Actually this is the same way as ExtraAttributes work, maybe ExtraAttribute workflow could be tweaked, so they can publish to a custom topic, and in plaintext not in json.

Or create some dummy HomeAssistantCommand which doesn't have a command topic, and connect the 2 sensors to that.

**Comment by richibrics** [comment]
Well I prefer the Command way but I have to figure it out the best approach, thank you

**Comment by richibrics** [comment]
@infeeeee I can use a simple sensor connected to a command, then I define the topic of latest version to be the extra attributes of current version and set the value template to extract the value from the extra attributes json

**Comment by richibrics** [comment]
It's still a work in progress but looks great to show the update in hass: 
image

**Comment by richibrics** [comment]
Need to come back to previous idea: having extra attributes 
- doesn't allow to have latest version in other WH
- needs a way to link the extra attribute topic dynamically

**Comment by richibrics** [comment]
HAss supports the Update entity discovery payload without command topic so to not show the "install" button. However without the Install button, the update doesn't show up in the available updates page.
Currently the command topic is provided but the script shows a warning message above the Install button that says the Install process must be run manually

**Comment by richibrics** [comment]
⚠️ A fake latest version is currently returned to show how the update looks like

**Review by infeeeee — COMMENTED** [review]
Very good!

The NO_REMOTE_INSTALL_AVAILABLE_MSG is cut for me on the HA frontend for some reason, maybe there is some hardcoded character limit?

kép

It really nice that HA translates everything correctly :)

**Comment by richibrics** [comment]
That's strange it doesn't happen to me

**Comment by infeeeee** [comment]
It cuts after exactly 255 characters, it's documented here: https://www.home-assistant.io/integrations/update.mqtt/#release_summary

> release_summary string (Optional)
>
> Summary of the release notes or changelog. This is suitable a brief update description of max 255 characters.

But this should be generated from the changelog in the future, and the `App.getDescription()` is not necessary there.

It's this line: https://github.com/richibrics/IoTuring/blob/94f575caae43b7b752d22985a53a6f39c5b526e6/IoTuring/Entity/Deployments/AppInfo/AppInfo.py#L80

**Comment by richibrics** [comment]
Okay thanks I remove the description from there

**Comment by infeeeee** [comment]
Wait, this secondary sensor not even needed. I read the docs again, you can just publish a json to the `state_topic`, and `latest_version_topic` is not needed in that case, so only one sensor for one optional command:

> JSON can also be used as state_topic payload.


```
{
  "installed_version": "1.21.0",
  "latest_version": "1.22.0",
  "title": "Device Firmware",
  "release_url": "https://example.com/release",
  "release_summary": "A new version of our amazing firmware",
  "entity_picture": "https://example.com/icon.png"
```

At the bottom of the page: https://www.home-assistant.io/integrations/update.mqtt/#examples

**Comment by richibrics** [comment]
Yes I know but in that way we would have lost the latest version information in other warehouses. Also we would have needed to add to the discovery payload another attibute with the same topic as the state_topic: since the state_topic is dynamically computed and can't be set neither in entities.yaml or in AppInfo custom payload method, we would have had to do some other changes in the Hass WH.
WIth secondary sensor now we are able to support also other types of HomeAssistant entities that need additional attributes topic in the payload

**Comment by richibrics** [comment]
Are you sure the latest version topic automatically sets to the state topic ? 
In the Hass example they set it in both
```

# Example configuration.yaml entry
mqtt:
  - update:
      name: "Shelly Plug S Firmware Update"
      title: "Shelly Plug S Firmware"
      release_url: "https://shelly-api-docs.shelly.cloud/gen1/#changelog"
      entity_picture: "https://brands.home-assistant.io/_/shelly/icon.png"
      state_topic: "shellies/shellyplug-s-112233/info"
      value_template: "{{ value_json['update'].old_version }}"
      latest_version_topic: "shellies/shellyplug-s-112233/info"
      latest_version_template: "{% if value_json['update'].new_version %}{{ value_json['update'].new_version }}{% else %}{{ value_json['update'].old_version }}{% endif %}"
      device_class: "firmware"
      command_topic: "shellies/shellyplug-s-112233/command"
      payload_install: "update_fw"
```

**Comment by richibrics** [comment]
@infeeeee what do you think ? I'd leave it as is to make it available to other WH too

**Comment by infeeeee** [comment]
Both ways have pros and cons:

Pros of this workflow:
- MQTT and console warehouse can easily get the current and new version numbers

Cons of this workflow:
- Data in the custom payload is not available in other warehouses only in HA, e.g: `release_url`, `release_summary`
- `release_summary` should be updated the same time as new version number, now it updates only after IoTuring restarts. So if there is a new update, the summary is not updated even in HA, users won't get info about new changes, it should be sent during Update, so sending discovery should be changed as well: https://www.home-assistant.io/integrations/update.mqtt/#release_summary 
- Technical debt: I don't know if we will ever need multiple sensors for any other entities, having this big workaround only for this single usecase seems a bit overkill

What should be changed if we send everything as a json/dict from a single sensor:
- Console warehouse: some valueformatter to display the sensor value as it's a dict not a single string, but every value is just a string in the dict, so it shouldn't be that complex
- MQTT Warehouse: I don't know how people use the MQTTWarehouse, but I guess it's for advanced users, so they can parse the json and get the data they want from there

Comparing pros and cons, I think sending everything in a json in a single sensor should be better solution.

**Comment by infeeeee** [comment]
I was thinking about this, and I found more pros for this workflow:

Maybe this multi sensor workflow would be useful elsewhere. MQTT Cover can send position and tilt values in a separate topics similarly: https://www.home-assistant.io/integrations/cover.mqtt/#position_topic But this needs 2 more sensors.

Maybe the ExtraAttributes could become a separate sensor as well, converted to ExtraAttribute json only in HA WH, their data could be visible in other WHs, as ExtraAtrributes now only visible in HA.

To solve the `release_summary` problem, the Entity should notify HA WH some way to update `discovery_payload`. `EntityData.GetCustomPayload()` already exists, with something like that.

So now I think this workflow would be better :)

**Comment by richibrics** [comment]
I like this change of idea :)

**Comment by richibrics** [comment]
@infeeeee How to place the release summary in just 255 chars ? I'd leave for sure the NO_REMOTE_INSTALL_AVAILABLE_MSG and then the space is really small

**Comment by richibrics** [comment]
Another thing, currently the custom payload is set at command register and not updated later

**Comment by richibrics** [comment]
Currently it's trimmed to place "..." when additional release summary entries don't fit in 255 chars
